src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -62,111 +61,115 @@
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
-   refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+   old-instances :- [JRubyInstance]
+   refill? :- schema/Bool
+   on-complete :- IDeref]
+  (let [pool (jruby-internal/get-pool pool-context)
+        pool-size (jruby-internal/get-pool-size pool-context)
+        new-instance-ids (map inc (range pool-size))
+        config (:config pool-context)
         cleanup-fn (get-in config [:lifecycle :cleanup])]
-    (log/info "Replacing old JRuby pool with new instance.")
-    (reset! pool-state-atom new-pool-state)
-    (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range old-pool-size)]
+    (doseq [[old-instance new-id] (zipmap old-instances new-instance-ids)]
       (try
-        (let [id (inc i)
-              instance (jruby-internal/borrow-from-pool!*
-                        jruby-internal/borrow-without-timeout-fn
-                        old-pool)]
-          (try
-            (jruby-internal/cleanup-pool-instance! instance cleanup-fn)
-            (when refill?
-              (jruby-internal/create-pool-instance! new-pool id config
-                                                    (partial send-flush-instance! pool-context))
-              (log/infof "Finished creating JRubyInstance %d of %d"
-                         id old-pool-size))
-            (finally
-              (.releaseItem old-pool instance false))))
+        (jruby-internal/cleanup-pool-instance! old-instance cleanup-fn)
+        (when refill?
+          (jruby-internal/create-pool-instance! pool new-id config
+                                                (partial send-flush-instance! pool-context))
+          (log/infof "Finished creating JRubyInstance %d of %d"
+                     new-id pool-size))
         (catch Exception e
-          (.clear new-pool)
-          (.insertPill new-pool (PoisonPill. e))
+          (.clear pool)
+          (jruby-internal/insert-poison-pill pool e)
           (throw (IllegalStateException.
-                  "There was a problem adding a JRubyInstance to the pool."
-                  e)))))
-    ;; Add a "RetryPoisonPill" to the pool in case something else is in the
-    ;; process of borrowing from the old pool.
-    (.insertPill old-pool (RetryPoisonPill. old-pool))))
+                  "There was a problem creating a JRubyInstance for the pool."
+                  e))))))
+  (if refill?
+    (log/info "Finished draining and refilling pool.")
+    (log/info "Finished draining pool."))
+  (deliver on-complete true))
```

Comment:
We should make sure that the `on-complete` promise is delivered even if an exception occurs during the cleanup/creation loop above, not just if we got all the way through the loop successfully.  Since the only function which calls `cleanup-and-refill-pool` appears to be `drain-and-refill-pool!`, I wonder if it would be better to set the promise from within the call to `cleanup-and-refill-pool` instead, and not bother passing the promise in here at all.

So, maybe in `drain-and-refill-pool!`, we could instead have:

```clj
  (send-agent (get-modify-instance-agent pool-context)
                 #(try
                      (cleanup-and-refill-pool pool-context old-instances refill?)
                      (finally
                           (deliver on-complete true))
```